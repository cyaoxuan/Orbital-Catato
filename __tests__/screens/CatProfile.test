import { cleanup, fireEvent, render } from "@testing-library/react-native";
import {
    CatProfile,
    CatProfileScreen,
} from "../../app/screens/main/catalogue/CatProfile";

afterEach(cleanup);

// Set time
beforeAll(() => {
    jest.useFakeTimers("modern");
    jest.setSystemTime(new Date(2023, 5, 20, 13, 12, 0, 0));
});

afterAll(() => {
    jest.useRealTimers();
});

// Mock getCat(catID)
jest.mock("../../app/screens/main/catalogue/CatProfile", () => {
    const originalModule = jest.requireActual(
        "../../app/screens/main/catalogue/CatProfile"
    );

    const cats = [
        {
            catID: 1,
            name: "Kitty",
            photoURLs: [
                require("../../assets/cats/cat-1-1.jpg"),
                require("../../assets/cats/cat-1-2.jpg"),
                require("../../assets/cats/cat-1-3.jpg"),
                require("../../assets/cats/cat-1-4.jpg"),
                require("../../assets/cats/cat-1-5.jpg"),
                require("../../assets/cats/cat-1-6.jpg"),
                require("../../assets/cats/cat-1-7.jpg"),
            ],
            gender: "M",
            birthYear: 2019,
            sterilised: true,
            keyFeatures: "Grey tabby and loves to sleep on his back",
            lastSeenLocation: "Engineering", // Will eventually be a geohash
            lastSeenTime: new Date(2023, 4, 19, 14, 21, 0, 0), // Will be converted from firebase timestamp
            lastFedTime: new Date(2023, 4, 19, 12, 1, 0, 0),
            concernStatus: ["Injured"],
            concernDesc: "Seen limping, possibly right front leg injured",
        },
    ];

    const mockGetCat = jest.fn((catID) =>
        cats.find((cat) => cat.catID === catID)
    );
    return {
        ...originalModule,
        getCat: mockGetCat,
    };
});

// Mock useNavigation.navigate
const mockNavigate = jest.fn();

jest.mock("expo-router", () => ({
    useNavigation: () => ({
        navigate: mockNavigate,
    }),
}));

// Clearing all mocks because navigate is called twice
afterEach(() => {
    jest.clearAllMocks();
});

// Mock useRoute
jest.mock("@react-navigation/native", () => ({
    useRoute: jest.fn(() => ({
        params: {
            catID: 1,
        },
    })),
}));

describe("<CatProfileScreen />", () => {
    it("renders CatProfileScreen successfully", () => {
        const profileContainer = render(<CatProfileScreen />);
        expect(profileContainer).toBeDefined();
    });

    it("renders 'Cat not found' message if cat is not found", () => {
        const { getByText } = render(<CatProfileScreen catID={6} />);
        const notFoundMessage = getByText("Cat not found: 6");
        expect(notFoundMessage).toBeDefined();
    });

    it("renders CatProfileScreen with cat correctly", () => {
        const tree = render(<CatProfileScreen catID={1} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("navigates to PhotoGallery when 'PhotosContainer' is pressed", () => {
        const { getByTestId } = render(<CatProfileScreen catID={1} />);
        const galleryButton = getByTestId("view-button");

        fireEvent.press(galleryButton);

        expect(mockNavigate).toHaveBeenCalledWith("PhotoGallery", {
            catID: 1,
            name: "Kitty",
            photoURLs: [
                require("../../assets/cats/cat-1-1.jpg"),
                require("../../assets/cats/cat-1-2.jpg"),
                require("../../assets/cats/cat-1-3.jpg"),
                require("../../assets/cats/cat-1-4.jpg"),
                require("../../assets/cats/cat-1-5.jpg"),
                require("../../assets/cats/cat-1-6.jpg"),
                require("../../assets/cats/cat-1-7.jpg"),
            ],
        });
    });

    it("navigates to Update Cat screen when 'Update Cat' button is pressed", () => {
        const { getByTestId } = render(<CatProfileScreen catID={1} />);
        const buttonContainerOuterLayer = getByTestId(
            "button-container-outer-layer"
        );
        const buttonContainer = buttonContainerOuterLayer.findByProps({
            testID: "button-container",
        });
        const button = buttonContainer.findByProps({ testID: "button" });
        fireEvent.press(button);

        expect(mockNavigate).toHaveBeenCalledWith("update", {
            screen: "Update",
            params: {
                cat: {
                    catID: 1,
                    name: "Kitty",
                    photoURLs: [
                        require("../../assets/cats/cat-1-1.jpg"),
                        require("../../assets/cats/cat-1-2.jpg"),
                        require("../../assets/cats/cat-1-3.jpg"),
                        require("../../assets/cats/cat-1-4.jpg"),
                        require("../../assets/cats/cat-1-5.jpg"),
                        require("../../assets/cats/cat-1-6.jpg"),
                        require("../../assets/cats/cat-1-7.jpg"),
                    ],
                    gender: "M",
                    birthYear: 2019,
                    sterilised: true,
                    keyFeatures: "Grey tabby and loves to sleep on his back",
                    concernStatus: ["Injured"],
                    concernDesc:
                        "Seen limping, possibly right front leg injured",
                },
            },
        });
    });
});
