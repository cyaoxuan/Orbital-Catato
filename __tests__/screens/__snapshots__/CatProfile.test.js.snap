// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatProfileScreen /> renders CatProfileScreen with cat correctly 1`] = `
<RCTScrollView
  testID="profile-container"
>
  <View>
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "padding": 15,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
          }
        }
      >
        <View
          style={
            [
              {
                "backgroundColor": "rgba(103, 80, 164, 1)",
                "borderRadius": 100,
                "height": 200,
                "width": 200,
              },
              undefined,
            ]
          }
        >
          <Image
            accessibilityIgnoresInvertColors={true}
            source={1}
            style={
              {
                "borderRadius": 100,
                "height": 200,
                "width": 200,
              }
            }
            testID="avatar-image"
          />
        </View>
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 32,
                "fontWeight": "400",
                "letterSpacing": 0,
                "lineHeight": 40,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              undefined,
            ]
          }
        >
          Kitty
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "backgroundColor": "lightgrey",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "padding": 15,
        }
      }
    >
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "fontWeight": "bold",
              },
            ]
          }
        >
          Gender
        </Text>
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              undefined,
            ]
          }
        >
          M
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "fontWeight": "bold",
              },
            ]
          }
        >
          Age
        </Text>
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              undefined,
            ]
          }
        >
          4y
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "justifyContent": "center",
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "fontWeight": "bold",
              },
            ]
          }
        >
          Sterilised
        </Text>
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              undefined,
            ]
          }
        >
          Yes
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "justifyContent": "space-evenly",
          "padding": 15,
        }
      }
    >
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "margin": 8,
            },
          ]
        }
      >
        <Text />
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Key Features: 
          <Text
            style={
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                {
                  "flex": 1,
                },
              ]
            }
          >
            Grey tabby and loves to sleep on his back
          </Text>
        </Text>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "margin": 8,
            },
          ]
        }
      >
        <Text />
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Last Seen: 
          <Text
            style={
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                {
                  "flex": 1,
                },
              ]
            }
          >
            Engineering, 19/05, 22:21 (31d ago)
          </Text>
        </Text>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "margin": 8,
            },
          ]
        }
      >
        <Text />
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Last Fed: 
          <Text
            style={
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                {
                  "flex": 1,
                },
              ]
            }
          >
            19/05, 20:01 (32d ago)
          </Text>
        </Text>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "margin": 8,
            },
          ]
        }
      >
        <Text />
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Concern Status: 
          <Text
            style={
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                {
                  "flex": 1,
                },
              ]
            }
          >
            Injured
          </Text>
        </Text>
      </View>
      <View
        style={
          [
            {
              "flexDirection": "row",
            },
            {
              "margin": 8,
            },
          ]
        }
      >
        <Text />
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 20,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              {
                "flex": 1,
                "fontWeight": "bold",
              },
            ]
          }
        >
          Concerns: 
          <Text
            style={
              [
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                },
                {
                  "textAlign": "left",
                },
                {
                  "color": "rgba(28, 27, 31, 1)",
                  "writingDirection": "ltr",
                },
                {
                  "flex": 1,
                },
              ]
            }
          >
            Seen limping, possibly right front leg injured
          </Text>
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "padding": 16,
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "space-between",
            "paddingHorizontal": 8,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "400",
                "letterSpacing": 0.15,
                "lineHeight": 24,
              },
              {
                "textAlign": "left",
              },
              {
                "color": "rgba(28, 27, 31, 1)",
                "writingDirection": "ltr",
              },
              undefined,
            ]
          }
        >
          Photos
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
          testID="view-button"
        >
          <View
            style={
              {
                "flexDirection": "row",
              }
            }
          >
            <Text
              style={
                [
                  {
                    "fontFamily": "System",
                    "fontSize": 16,
                    "fontWeight": "400",
                    "letterSpacing": 0.15,
                    "lineHeight": 24,
                  },
                  {
                    "textAlign": "left",
                  },
                  {
                    "color": "rgba(28, 27, 31, 1)",
                    "writingDirection": "ltr",
                  },
                  undefined,
                ]
              }
            >
              View All
            </Text>
            <Text />
          </View>
        </View>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "rgba(202, 196, 208, 1)",
              "height": 0.5,
            },
            undefined,
            false,
            false,
            undefined,
          ]
        }
      />
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          testID="preview-photo"
        >
          <Image
            source={1}
            style={
              {
                "height": 171.5,
                "margin": 4,
                "resizeMode": "cover",
                "width": 171.5,
              }
            }
          />
        </View>
        <View
          testID="preview-photo"
        >
          <Image
            source={1}
            style={
              {
                "height": 171.5,
                "margin": 4,
                "resizeMode": "cover",
                "width": 171.5,
              }
            }
          />
        </View>
        <View
          testID="preview-photo"
        >
          <Image
            source={1}
            style={
              {
                "height": 171.5,
                "margin": 4,
                "resizeMode": "cover",
                "width": 171.5,
              }
            }
          />
        </View>
        <View
          testID="preview-photo"
        >
          <Image
            source={1}
            style={
              {
                "height": 171.5,
                "margin": 4,
                "resizeMode": "cover",
                "width": 171.5,
              }
            }
          />
        </View>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
        }
      }
    >
      <View
        style={
          {
            "flexDirection": "row",
          }
        }
      >
        <View
          collapsable={false}
          style={
            {
              "backgroundColor": "transparent",
              "borderRadius": 20,
              "margin": 10,
              "shadowColor": "#000",
              "shadowOffset": {
                "height": 0,
                "width": 0,
              },
              "shadowOpacity": 0,
              "shadowRadius": 0,
              "width": "60%",
            }
          }
          testID="button-container-outer-layer"
        >
          <View
            collapsable={false}
            style={
              {
                "backgroundColor": "transparent",
                "borderColor": "rgba(121, 116, 126, 1)",
                "borderRadius": 20,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flex": undefined,
                "minWidth": 64,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
              }
            }
            testID="button-container"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": undefined,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                [
                  {
                    "overflow": "hidden",
                  },
                  {
                    "borderRadius": 20,
                  },
                ]
              }
              testID="button"
            >
              <View
                style={
                  [
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  selectable={false}
                  style={
                    [
                      {
                        "fontFamily": "System",
                        "fontSize": 14,
                        "fontWeight": "500",
                        "letterSpacing": 0.1,
                        "lineHeight": 20,
                      },
                      {
                        "textAlign": "left",
                      },
                      {
                        "color": "rgba(28, 27, 31, 1)",
                        "writingDirection": "ltr",
                      },
                      [
                        {
                          "marginHorizontal": 16,
                          "marginVertical": 9,
                          "textAlign": "center",
                        },
                        false,
                        {
                          "marginHorizontal": 24,
                          "marginVertical": 10,
                        },
                        undefined,
                        false,
                        {
                          "color": "rgba(103, 80, 164, 1)",
                          "fontFamily": "System",
                          "fontSize": 14,
                          "fontWeight": "500",
                          "letterSpacing": 0.1,
                          "lineHeight": 20,
                        },
                        undefined,
                      ],
                    ]
                  }
                  testID="button-text"
                >
                  Update Cat
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
